### PR TITLE
add precommit

### PR DESCRIPTION
**Pull request by Borda**
### Description

Adding pre-commit bot, need to be installed https://github.com/apps/pre-commit-ci/installations/new

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [x] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [x] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Review by edgarriba — APPROVED** [review]
LGTM. Do we have to make a first run with yapf to format the code ?

**Comment by Borda** [comment]
lets wait till #1039 and #1040 are resolved :]

**Review by edgarriba — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      - id: check-yaml
+      - id: pretty-format-json
+
+  - repo: https://github.com/pre-commit/mirrors-yapf
+    rev: v0.31.0
+    hooks:
+      - id: yapf
+        name: formatting
+        entry: yapf
+        args: [--parallel]
```
Comment (asottile) [inline .pre-commit-config.yaml:24]: by the way, you're fork bombing -- pre-commit runs things in parallel by default (and tends to do a better job than `yapf` does)

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        language: python
+        types: [python]
```
Comment (asottile) [inline .pre-commit-config.yaml:26]: these lines are unnecessary -- they come from the base tool

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      - id: trailing-whitespace
+      - id: check-yaml
+      - id: pretty-format-json
+
+  - repo: https://github.com/pre-commit/mirrors-yapf
+    rev: v0.31.0
+    hooks:
+      - id: yapf
+        name: formatting
+        entry: yapf
```
Comment (asottile) [inline .pre-commit-config.yaml:23]: this line is unnecessary, it comes from the base tool

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        types: [python]
+
+  - repo: https://github.com/PyCQA/flake8
+    rev: 3.9.2
+    hooks:
+      - id: flake8
+        name: PEP8
+        entry: flake8
+        language: python
+        types: [python]
```
Comment (asottile) [inline .pre-commit-config.yaml:35]: these 3 lines are unnecessary, they come from the base tool

**Comment by asottile** [comment]
(sorry for the drive-by, your runs tripped some alarms!)

I'm amazed at just how slow `yapf` is -- you may find `autopep8` / `add-trailing-comma` / `black` to be faster and more reliable

there are a handful of files which make `yapf` really really unhappy in this codebase (it doesn't do well with attempting to refold deeply nested things)

**Pull request merged**